src/Microsoft.ML.Transforms/WhiteningTransform.cs

```diff
@@ -383,249 +507,339 @@ private void TrainModels(Float[][] columnData, int[] rowCounts, IChannel ch)
+                _srcTypes = new ColumnType[_parent.ColumnPairs.Length];
 
-        protected override ColumnType GetColumnTypeCore(int iinfo)
-        {
-            Host.Check(0 <= iinfo & iinfo < Infos.Length);
-            return _exes[iinfo].Type;
-        }
+                for (int i = 0; i < _parent.ColumnPairs.Length; i++)
+                {
+                    var info = _parent._infos[i];
+                    if (!InputSchema.TryGetColumnIndex(_parent.ColumnPairs[i].input, out _cols[i]))
+                        throw Host.ExceptSchemaMismatch(nameof(inputSchema), "input", _parent.ColumnPairs[i].input);
+                    _srcTypes[i] = inputSchema.GetColumnType(_cols[i]);
+                    ColumnType outType = (info.Kind == WhiteningKind.Pca && info.PcaNum > 0) ? new VectorType(NumberType.Float, info.PcaNum) : _srcTypes[i];
+                    ValidateModel(Host, _parent._models[i], outType);
+                    if (info.SaveInv)
+                        ValidateModel(Host, _parent._invModels[i], outType);
+                }
+            }
 
-        protected override Delegate GetGetterCore(IChannel ch, IRow input, int iinfo, out Action disposer)
-        {
-            Host.AssertValueOrNull(ch);
-            Host.AssertValue(input);
-            Host.Assert(0 <= iinfo && iinfo < Infos.Length);
-            disposer = null;
-
-            var ex = _exes[iinfo];
-            Host.Assert(ex.Kind == WhiteningKind.Pca || ex.Kind == WhiteningKind.Zca);
-            var getSrc = GetSrcGetter<VBuffer<Float>>(input, iinfo);
-            var src = default(VBuffer<Float>);
-            int cslotSrc = Infos[iinfo].TypeSrc.ValueCount;
-            int cslotDst = (ex.Kind == WhiteningKind.Pca && ex.PcaNum > 0) ? ex.PcaNum : Infos[iinfo].TypeSrc.ValueCount;
-            var model = _models[iinfo];
-            ValueGetter<VBuffer<Float>> del =
-                (ref VBuffer<Float> dst) =>
+            public override Schema.Column[] GetOutputColumns()
+            {
+                var result = new Schema.Column[_parent.ColumnPairs.Length];
+                for (int iinfo = 0; iinfo < _parent.ColumnPairs.Length; iinfo++)
                 {
-                    getSrc(ref src);
-                    Host.Check(src.Length == cslotSrc, "Invalid column size.");
-                    FillValues(model, ref src, ref dst, cslotDst);
-                };
-            return del;
-        }
+                    InputSchema.TryGetColumnIndex(_parent.ColumnPairs[iinfo].input, out int colIndex);
+                    Host.Assert(colIndex >= 0);
+                    var info = _parent._infos[iinfo];
+                    // For ZCA the output type is the same as the input type. For PCA the size of the output vector is determined by PCANum.
+                    ColumnType outType = (info.Kind == WhiteningKind.Pca && info.PcaNum > 0) ? new VectorType(NumberType.Float, info.PcaNum) : _srcTypes[iinfo];
```

Comment:
What I was expecting is something like: For PCA, the transform equation is y=U^Tx, where "^T" denotes matrix transpose, x is an 1-D vector (i.e., the input column), and U=[u_1, ..., u_PcaNum] is a n-by-PcaNum matrix. The symbol u_k is the k-th largest (in terms of the associated eigenvalue) eigenvector of \sum_{i=1}^m x_ix_i^T, where x_i is the whitened column at the i-th row and we have m rows in the training data.

Reply:
I agree that it would be useful to have a more clear explanation of the algorithm in a comment but I am not sure this should go in GetOutputColumns methods, as this mostly deals with the shape, and type of the output.

---
In reply to: [229773103](https://github.com/dotnet/machinelearning/pull/1452#discussion_r229773103) [](ancestors = 229773103)

Reply:
I believe so. How could user know what this function is doing without those details? Maybe this function looks simple, but it produces the most important information for IR.

---
In reply to: [230214642](https://github.com/dotnet/machinelearning/pull/1452#discussion_r230214642) [](ancestors = 230214642,229773103)

Reply:
Ok I agree actually. You are right! I am adding it as a comment and will do the same for zca.

---
In reply to: [230504546](https://github.com/dotnet/machinelearning/pull/1452#discussion_r230504546) [](ancestors = 230504546,230214642,229773103)

Reply:
Many thanks for making this comment. Could you make it < summary > … < /summary > of this public function?